pkg/testcase/upgradenodereplacement.go

```diff
@@ -0,0 +1,570 @@
+	}
+
+	// delete the last remaining server = leader
+	if delErr := deleteServer(serverLeaderIp, awsDependencies); delErr != nil {
+		return delErr
+	}
+	shared.LogLevel("info", "Last Server deleted ip: %s\n", serverLeaderIp)
+
+	return nil
+}
+
+func scpToNewNodes(product, nodeType string, newNodeIps []string) error {
+	if newNodeIps == nil {
+		return shared.ReturnLogError("newServerIps should send at least one ip\n")
+	}
+
+	if product != "k3s" && product != "rke2" {
+		return shared.ReturnLogError("unsupported product: %s\n", product)
+	}
+
+	chanErr := make(chan error, len(newNodeIps))
+	var wg sync.WaitGroup
+
+	for _, ip := range newNodeIps {
+		wg.Add(1)
+		go func(ip string) {
+			defer wg.Done()
+			var err error
+			if product == "k3s" {
+				err = scpK3sFiles(product, nodeType, ip)
+			} else {
+				err = scpRke2Files(product, nodeType, ip)
+			}
+			if err != nil {
+				chanErr <- shared.ReturnLogError("error scp files to new nodes: %w\n", err)
+				close(chanErr)
+			}
+		}(ip)
+	}
+
+	wg.Wait()
+	close(chanErr)
+
+	return nil
+}
+
+func scpRke2Files(product, nodeType, ip string) error {
+	joinLocalPath := shared.BasePath() + fmt.Sprintf("/modules/install/join_rke2_%s.sh", nodeType)
+	joinRemotePath := fmt.Sprintf("/tmp/join_rke2_%s.sh", nodeType)
+
+	if err := shared.RunScp(ip, product, []string{joinLocalPath}, []string{joinRemotePath}); err != nil {
+		return shared.ReturnLogError("error running scp: %w with ip: %s", err, ip)
+	}
+
+	return nil
+}
+
+func scpK3sFiles(product, nodeType, ip string) error {
+	if nodeType == agent {
+		cisWorkerLocalPath := shared.BasePath() + "/modules/k3s/worker/cis_worker_config.yaml"
+		cisWorkerRemotePath := "/tmp/cis_worker_config.yaml"
+
+		joinLocalPath := shared.BasePath() + fmt.Sprintf("/modules/install/join_k3s_%s.sh", agent)
+		joinRemotePath := fmt.Sprintf("/tmp/join_k3s_%s.sh", agent)
+
+		if err := shared.RunScp(
+			ip,
+			product,
+			[]string{cisWorkerLocalPath, joinLocalPath},
+			[]string{cisWorkerRemotePath, joinRemotePath},
+		); err != nil {
+			return err
+		}
+	} else {
+		cisMasterLocalPath := shared.BasePath() + "/modules/k3s/master/cis_master_config.yaml"
+		cisMasterRemotePath := "/tmp/cis_master_config.yaml"
+
+		clusterLevelpssLocalPath := shared.BasePath() + "/modules/k3s/master/cluster-level-pss.yaml"
+		clusterLevelpssRemotePath := "/tmp/cluster-level-pss.yaml"
+
+		auditLocalPath := shared.BasePath() + "/modules/k3s/master/audit.yaml"
+		auditRemotePath := "/tmp/audit.yaml"
+
+		policyLocalPath := shared.BasePath() + "/modules/k3s/master/policy.yaml"
+		policyRemotePath := "/tmp/policy.yaml"
+
+		ingressPolicyLocalPath := shared.BasePath() + "/modules/k3s/master/ingresspolicy.yaml"
+		ingressPolicyRemotePath := "/tmp/ingresspolicy.yaml"
+
+		joinLocalPath := shared.BasePath() + fmt.Sprintf("/modules/install/join_k3s_%s.sh", master)
+		joinRemotePath := fmt.Sprintf("/tmp/join_k3s_%s.sh", master)
```

Comment:
We should probably create an item in the backlog to cleanup all the hardening of k3s. This is complicated and it's because of all the separate files, when it could probably be just one file now. Nothing to change here on this PR, just something I noticed because of it!

Reply:
agreed

Reply:
https://github.com/rancher/qa-tasks/issues/1224